packages/react-server-cli/src/coreJsMiddleware.js

```diff
@@ -1,17 +1,33 @@
-export default (pathToStatic) => {
+export default (pathToStatic, manifest) => {
 	return class CoreJsMiddleware {
 		getSystemScripts(next) {
 			const routeName = this.getRequest().getRouteName();
 			const baseUrl = pathToStatic || "/";
+
+			const scripts = [];
+
+			// read out the chunkmap so that the client code knows how to download the
+			// other chunks.
+			scripts.push({
+				type: "text/javascript",
+				text: `window.webpackManifest = ${JSON.stringify(manifest.jsChunksById)};`,
+			});
+
+			scripts.push(`${baseUrl}${manifest.jsChunksByName.common}`);
+			scripts.push(`${baseUrl}${manifest.jsChunksByName[routeName]}`);
+
 			return [
-				`${baseUrl}common.js`,
-				`${baseUrl}${routeName}.bundle.js`,
+				...scripts,
 				{
 					type: "text/javascript",
-					text: baseUrl ? `
-						if (typeof window !== "undefined" && window.__setReactServerBase) {
-							window.__setReactServerBase(${JSON.stringify(baseUrl)});
-						}` : "",
+					text: `
+						if (typeof window !== "undefined") {
+							if (window.__setReactServerBase) {
+								window.__setReactServerBase(${JSON.stringify(baseUrl)});
+							}
+							window.__reactServerManifest = ${manifest ? JSON.stringify(manifest) : "{}"};
```

Comment:
It could get big; it's probably ~100 bytes per entry point (that's a guess; I haven't looked at it to measure).

I believe the issue with it being in the bundle is that putting it in the bundle would change all bundle hashes any time any file changes and defeat the purpose of long-term caching.